Dozuki/CloudPrem-Infra
# Terraform Refactor to Physical & Logical Modules

# Background
What we're seeing with the 4 module approach is greater flexibility for development, being able to spin up and down just a small subset of our terraform but when It comes to end-user code pipeline use it creates quite a bit of added complexity, especially when handling partial creation failures. In the old stack if you started to spin up an environment and it failed on the storage or compute module but network was already successfully created, there was no way to delete the stack without manual workstation intervention because Terragrunt would always try to destroy the stack in reverse order (app, storage, compute, network) and if those modules weren't created yet, Terragrunt could not handle that and would just die. Additionally with 4 modules you had 4 greater opportunities at failure in both directions: create and destroy. 

# Solution
It turns out a modular approach was a good idea but it makes more sense to have fewer modules that bifurcate using different logic. We now have 2 modules: Physical - which includes all physical resources, like the k8 cluster, databases, ec2 instances, IAM roles & policies, etc, and Logical - which is all the resources that are deployed _onto_ the physical resources. So k8 manifests & helm charts mostly. Not only does this modular configuration reduce our error possibility but it makes environment deletion a snap. We only have to run the destroy command on the physical module since deleting the physical stuff will automatically delete all the logical resources deployed onto them. This makes for a much more reliable code pipeline as well. If something breaks somewhere, the deletion phase is only 1 module long so it will always be able to clean up after itself regardless of what might fail during creation. 

# QA
The live folder structure has changed to support the dual module setup. It should be self explanatory what each of these environments is for but for more info checkout the README.md file inside the live folder.

To spin up this new terraform code for QA you would pick a region and an account first, let's say standard account and us-east-1 region, then follow the steps below:

1. If using our testing AWS account and the default local awscli profile, you don't need to modify the account.hcl file anymore.
2. cd into the environment you want to test, let's say `live/standard/us-east-1/min` for this example.
3. Run terragrunt to apply the initial stack (plan isn't supported for a new stack): `terragrunt run-all apply` and type `y` to confirm.
4. Go get a coffee

Once complete you'll be presented with 3 outputs: the replicated dashboard URL, the dashboard password, and the application url. Visit the app url and the dozuki login page should be present (you may need to wait 5 minutes or so after a fresh deploy for the app to finish starting up, if it takes longer than that login to the dashboard and investigate)


# Notable Changes
Along with the main terragrunt refactor I've made many upgrades to the system as a whole:

1. Refactored S3 buckets to no longer use the terraform s3 module
   * This module was causing all kinds of issues with tfsec and honestly was way more complicated than we needed for simple buckets.
   * S3 buckets no longer allow public ACL's of any kind to follow AWS S3 best security practices.
   * We now make use of a logging bucket to log all S3 changes in an environment.
   * [Source for above](https://github.com/Dozuki/CloudPrem-Infra/blob/physical-logical/cloudprem/physical/s3.tf)
2.  Massive refactor of the DMS system for BI environments.
    * We wait until the replicated app is fully available before starting the DMS replication task. This ensures our entire database gets replicated on first start. For some reason DMS seems flakey when replicating database and table additions with our system. 
      * [source](https://github.com/Dozuki/CloudPrem-Infra/blob/physical-logical/cloudprem/logical/bi.tf)
    * We now use a utility shell script to stop the DMS task on destroy operations. Using the built-in null resource local-exec provisioner alone was not sufficient for this asynchronous process. The script now stops the task and waits for it to be fully stopped before allowing deletion to prevent errors 
      * [invocation](https://github.com/Dozuki/CloudPrem-Infra/blob/5f57a82e73ab986e6efc12606c66a4f6bff6f8db/cloudprem/physical/rds.tf#L331) and [source](https://github.com/Dozuki/CloudPrem-Infra/blob/physical-logical/cloudprem/physical/util/dms-stop.sh)
    * DMS now uses in-transit SSL for all communication between master and replica RDS instances. 
      * [source](https://github.com/Dozuki/CloudPrem-Infra/blob/5f57a82e73ab986e6efc12606c66a4f6bff6f8db/cloudprem/physical/rds.tf#L249)
    * Expanded DMS config to include data validation and error logging to CloudWatch. 
      * [source](https://github.com/Dozuki/CloudPrem-Infra/blob/physical-logical/cloudprem/physical/static/dms_config.json)
 3. All K8 secrets are now encrypted using KMS 
    * [source](https://github.com/Dozuki/CloudPrem-Infra/blob/5f57a82e73ab986e6efc12606c66a4f6bff6f8db/cloudprem/physical/eks.tf#L205)
 4. Upgraded many terraform modules and the AWS provider to the latest version available at the time.
 5. Linked the dashboard password resource to the NLB id. This forces a new password to be generated for the environment if physical-only deletion and full-stack recreation occurs. 
    * [source](https://github.com/Dozuki/CloudPrem-Infra/blob/5f57a82e73ab986e6efc12606c66a4f6bff6f8db/cloudprem/logical/replicated.tf#L9)

# Caveats
This refactor is incompatible with our existing code pipeline and CloudFormation systems and will need some minor changes to maintain a similar workflow. For that reason even if/when this PR is merged will not be able to "release" this refactor until those changes are done as well. Thankfully we use Gitops for our terraform versioning so merging this into master won't break any existing installs.
